File: tim_can_bac_n_cua_a.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Ham tim khoang cach ly va gia tri ban dau x0
double find_initial_interval(double a, int n) {
    if (n <= 0) {
        printf("Error: n must be a positive integer\n");
        exit(1);
    }

    double left, right, x0;

    // Truong hop n chan
    if (n % 2 == 0) {
        if (a <= 0) {
            printf("Invalid\n");
            exit(1);
        } else if (0 < a && a < 1) {
            double h = 1.0;
            while (1) {
                if (pow(h, n) < a && a < 1) {
                    left = h;
                    right = 1;
                    break;
                }
                h /= 2;
            }
            printf("Interval: (%lf, %lf)\n", left, right);
        } else { // a > 1
            int k = 1;
            while (1) {
                if (pow(k, n) < a && a < pow(k + 1, n)) {
                    left = k;
                    right = k + 1;
                    break;
                }
                k++;
            }
            printf("Interval: (%lf, %lf) - Case a > 1\n", left, right);
        }
        x0 = (left + right) / 2;
        return x0;
    }
    // Truong hop n le
    else {
        if (a > 0) {
            if (a < 1) {
                double h = 1.0;
                while (1) {
                    if (pow(h, n) < a && a < 1) {
                        left = h;
                        right = 1;
                        break;
                    }
                    h /= 2;
                }
                printf("Interval: (%lf, %lf)\n", left, right);
            } else { // a > 1
                int k = 1;
                while (1) {
                    if (pow(k, n) < a && a < pow(k + 1, n)) {
                        left = k;
                        right = k + 1;
                        break;
                    }
                    k++;
                }
                printf("Interval: (%lf, %lf)\n", left, right);
            }
            x0 = (left + right) / 2;
            return x0;
        } else { // a < 0
            double b = -a;
            if (0 < b && b < 1) {
                double h = 1.0;
                while (1) {
                    if (pow(h, n) < b && b < 1) {
                        left = h;
                        right = 1;
                        break;
                    }
                    h /= 2;
                }
                printf("Interval for nth root of %f: (%lf, %lf)\n", b, left, right);
            } else { // b > 1
                int k = 1;
                while (1) {
                    if (pow(k, n) < b && b < pow(k + 1, n)) {
                        left = k;
                        right = k + 1;
                        break;
                    }
                    k++;
                }
                printf("Interval for nth root of %lf: (%lf, %lf)\n", b, left, right);
            }
            x0 = (left + right) / 2;
            return x0;
        }
    }
}

// Ham tinh can bac n bang phuong phap Newton-Raphson
double newton_nth_root(double a, int n) {
    double x = find_initial_interval(a, n);
    double epsilon = 1e-6; // Sai so epsilon
    int i = 0;
    double b;

    // Neu a < 0 va n le, tinh b = -a
    if (a < 0 && n % 2 == 1) {
        b = -a;
    } else {
        b = a;
    }

    while (1) {
        // Giai phuong trinh x^n - b = 0 bang phuong phap Newton
        double f_x = pow(x, n) - b;         // f(x) = x^n - b
        double f_prime_x = n * pow(x, n - 1); // f'(x) = n * x^(n-1)
        double x_new = x - f_x / f_prime_x; // x_new = x - f(x)/f'(x)
        double error = fabs(x_new - x);     // Tinh sai so
        
        printf("Iteration %d: x = %.9lf, error = %.9e\n", i, x_new, error);
        
        if (error < epsilon) {
            // Neu a < 0 va n le thi tra ve -x_new
            if (a < 0 && n % 2 == 1) {
                return -x_new;
            } else {
                return x_new;
            }
        }
        x = x_new;
        i++;
    }
}

int main() {
    double a = -17;
    int n = 5;
    double result = newton_nth_root(a, n);
    printf("\nResult: nth root %d of %lf = %.9lf\n", n, a, result);
    return 0;
}

File: tim_can_bac_n_cua_a.py
def find_initial_interval(a, n):
    if n <= 0:
        raise ValueError("Invalid")
    # n chan
    if n % 2 == 0:
        if a <= 0:
            raise ValueError("Invalid")
        elif 0 < a < 1:
            h = 1.0
            while True:
                if h**n < a < 1:
                    left, right = h, 1
                    break
                h /= 2
            print(f"Interval: ({left}, {right}) - Case 0 < a < 1")
        else:  # a > 1
            k = 1
            while True:
                if k**n < a < (k + 1)**n:
                    left, right = k, k + 1
                    break
                k += 1
            print(f"Interval: ({left}, {right}) - Case a > 1")
        x0 = (left + right) / 2
        return x0
    
    # n le 
    else:
        if a > 0:
            if a < 1:
                h = 1.0
                while True:
                    if h**n < a < 1:
                        left, right = h, 1
                        break
                    h /= 2
                print(f"Interval: ({left}, {right})")
            else:  # a > 1
                k = 1
                while True:
                    if k**n < a < (k + 1)**n:
                        left, right = k, k + 1
                        break
                    k += 1
                print(f"Interval: ({left}, {right})")
            x0 = (left + right) / 2
            return x0
        else:  # a < 0
            b = -a 
            if 0 < b < 1:
                h = 1.0
                while True:
                    if h**n < b < 1:
                        left, right = h, 1
                        break
                    h /= 2
                print(f"Interval for nth root of {b}: ({left}, {right})")
            else:  # b > 1
                k = 1
                while True:
                    if k**n < b < (k + 1)**n:
                        left, right = k, k + 1
                        break
                    k += 1
                print(f"Interval for nth root of {b}: ({left}, {right})")
            x0 = (left + right) / 2
            return x0

def newton_nth_root(a, n):
    x = find_initial_interval(a, n)
    epsilon = 1e-6  # Nhap sai so epsilon
    i = 0
    # Neu a < 0 va n le tinh b = -a
    if a < 0 and n % 2 == 1:
        b = -a
    else:
        b = a
    while True:
        # Giai phuong trinh x^n - b = 0 bang phuong phap Newton
        # f(x) = x^n - b
        f_x = x**n - b
        f_prime_x = n * x**(n - 1)
        x_new = x - f_x / f_prime_x
        error = abs(x_new - x)
        print(f"Iteration {i}: x = {x_new}, error = {error}")
        if error < epsilon:
            # neu a < 0 va n le thi tra ve -x_new
            if a < 0 and n % 2 == 1:
                return -x_new
            else:
                return x_new
        x = x_new
        i += 1

def main():
    a = -17 
    n = 5  
    
    try:
        result = newton_nth_root(a, n)
        print(f"\nResult: nth root {n} of {a} = {result}")
    except ValueError as e:
        print(f"Error with a = {a}, n = {n}: {e}")

if __name__ == "__main__":
    main()